numba/runtime/nrt.py

```diff
@@ -16,7 +15,6 @@ class _Runtime(object):
     def __init__(self):
         self._init = False
 
-    @lock_llvm # prevent race to install compiled library functions
```

Comment:
I think a lock here is still needed, it was an abuse of the already existing `@lock_llvm` for the purposes of ensuring threadsafety on initialisation opposed to threadsafety in LLVM calls themselves. I think we need a more general initialisation lock anyway as I'm fairly sure that this https://github.com/numba/numba/issues/2517 is a result of a race condition during the initialisation sequence, this lock should be applied here too. For ref, this patch added the current lock and has the explanation for the addition https://github.com/numba/numba/pull/2449. Thoughts?

Reply:
I can borrow the `numba.lock.lock_compiler` here.

Reply:
See https://github.com/numba/numba/pull/2627/commits/bdab4f1bae897b1770591a7b4fa0faf7a32610d5